### PR TITLE
ctxutil: better go 1.21 compatibility

### PR DESCRIPTION
**Pull request by knz**
Informs  #112088.
Needed for #111928.

Release note: None

### CODE REVIEW

**Comment by knz** [comment]
The following patch is needed in the go runtime to enable the 1.21 code:
```diff
diff --git a/src/context/context.go b/src/context/context.go
index ee66b43c85..5c451cf97f 100644
--- a/src/context/context.go
+++ b/src/context/context.go
@@ -423,6 +423,12 @@ type cancelCtx struct {
        cause    error                 // set to non-nil by the first cancel call
 }

+type ExportedCancelCtx = cancelCtx
+
+func (c *cancelCtx) ExportedPropagateCancel(parent Context, child canceler) {
+       c.propagateCancel(parent, child)
+}
+
 func (c *cancelCtx) Value(key any) any {
        if key == &cancelCtxKey {
                return c
```

**Review by miretskiy — APPROVED** [review]

File: pkg/util/ctxutil/canceler_1_21_bazel.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// A canceler is a context type that can be canceled directly. The
+// implementations are *cancelCtx and *timerCtx.
+// This interface definition is applicable to go1.20 or higher
+type canceler interface {
+	cancel(removeFromParent bool, err, cause error)
+	Done() <-chan struct{}
+}
+
+func (c *whenDone) cancel(removeFromParent bool, err, cause error) {
+	cx := (&c.ExportedCancelCtx).(canceler)
```
Comment (miretskiy) [inline pkg/util/ctxutil/canceler_1_21_bazel.go:26]: does this need &?

Comment (knz) [reply pkg/util/ctxutil/canceler_1_21_bazel.go:26]: `cancelCtx` implements canceler by pointer, not by value. So I guess yes.

**Comment by knz** [comment]
Apparently yevgueny's approach is better. so we don't have to proceed with this.

**Pull request closed without merging**